### ISSUE TITLE
[nightshift] Bus Factor Analysis

### ISSUE THREAD

**Issue by nightshift-micr** [Microck/ordinary-claude-skills#16]
# Bus Factor Analysis Report — ordinary-claude-skills

**Repository:** `~/nightshift-workspace/clones/ordinary-claude-skills`  
**Date:** 2026-04-14  
**Analyst:** Nightshift Code Quality Bot (Hermes Agent)  
**Scope:** Full codebase — 415 skill directories, 1,274 markdown files, 3032 categorized entries, 201 code files

---

## Executive Summary

The ordinary-claude-skills repository is a **monorepo aggregating 415+ Claude AI skill packages** from multiple upstream sources (Anthropic, ComposioHQ, K-Dense-AI, and community authors). It is curated by a **single maintainer (Microck)** with **no co-maintainers, no CI/CD, and no contribution guidelines at the repo level**.

The bus factor is **1**. If the sole maintainer becomes unavailable, there is no documented process, no automation, and no shared knowledge to continue operating this repository.

The most critical risks are:
- **P0:** Single maintainer with no backup; entire git history is 1 commit from 1 author
- **P1:** 4 complex multi-file skills (scientific-skills, win11-speckit, ios-simulator, notebooklm) totaling ~18MB and 360+ code files with domain-specific expertise
- **P1:** Undocumented duplication logic between `skills_all/` (415 dirs) and `skills_categorized/` (1444 dirs) — 58MB total with significant overlap

---

## Findings Table

| Severity | File/Area | Description | Recommendation |
|----------|-----------|-------------|----------------|
| **P0** | Entire repo | Single maintainer (1 author, 1 commit) — bus factor = 1 | Recruit 2+ co-maintainers; add CODEOWNERS |
| **P0** | No CONTRIBUTING.md at root | No contribution guidelines for the aggregation repo | Create CONTRIBUTING.md with skill submission process |
| **P1** | `skills_all/claude-scientific-skills/` | 14MB, 861 files, 144 Python scripts — massive domain-specific skill collection with zero internal docs on maintenance | Document the skill inventory, indexing process, and update cadence |
| **P1** | `skills_all/claude-win11-speckit-update-skill/` | 248 files, 37 PowerShell scripts (1059-line orchestrator) — Windows-specific domain with no PowerShell expertise documented | Add architecture doc; identify maintainers with PowerShell knowledge |
| **P1** | `skills_all/` ↔ `skills_categorized/` | 88 skills duplicated across 5+ categories; no documented mapping logic | Document categorization algorithm; consider generating `skills_categorized/` from `skills_all/` + config |
| **P1** | 15 skills in `skills_all/` | Missing from `skills_categorized/` — invisible to category-based navigation | Audit and fix categorization gaps |
| **P2** | `skills_all/*/metadata.json` | 15 of 415 skills lack metadata.json — inconsistent skill structure | Add metadata for all skills or auto-generate |
| **P2** | No CI/CD pipeline | Zero top-level GitHub Actions workflows | Add validation: check SKILL.md existence, metadata.json schema, dead links |
| **P2** | Categorization logic | 88 skills appear in >5 categories; some in 20 — no documented rationale for category assignments | Document category criteria; add categorization config file |
| **P2** | `skills_all/treatment-plans/SKILL.md` | 1536-line SKILL.md — complex medical domain knowledge with no modularization | Split into multiple reference files; add table of contents |
| **P2** | `skills_all/template-skill/SKILL.md` | 6-line stub with no actual template content | Flesh out template with standard sections and examples |
| **P3** | `skills_all/*/requirements.txt` | Only 3 of 415 skills have dependency files, yet many reference external tools | Standardize dependency declaration across all executable skills |
| **P3** | `docs/` static site | 880KB generated site with no documented build/deploy process | Document how to regenerate and deploy docs |
| **P3** | License ambiguity | Top-level MIT claims copyright to Microck, but most content is from upstream repos | Audit and document per-skill licensing clearly |

---

### P0-01: Single Maintainer — Bus Factor = 1

**Location:** Entire repository  
**Evidence:**
```
$ git shortlog -sn --all
1  Microck
```
The entire git history consists of **1 commit** (`0a7a18c`). All 415 skills, 1274 markdown files, and 201 code files were added in a single squashed merge. There are no branches, no CODEOWNERS file, and no co-maintainers documented.

**Impact:** If Microck becomes unavailable, nobody has documented context on:
- How skills are ingested from upstream sources
- How categorization decisions are made
- How the static docs site is built/deployed
- How to validate new skill submissions

**Recommendation:**
1. Add a `CODEOWNERS` file with at least 2 reviewers
2. Create `CONTRIBUTING.md` with skill submission guidelines
3. Document the maintenance workflow in a `MAINTAINERS.md`

---

### P0-02: No Contribution or Onboarding Documentation

**Location:** Repository root — missing `CONTRIBUTING.md`, `MAINTAINERS.md`, `CODEOWNERS`  
**Evidence:** While individual skills (e.g., `skills_all/image-enhancer/CONTRIBUTING.md`) have their own contributing guides carried over from upstream, the aggregation repo has none.

**Impact:** New contributors have no documented path to:
- Submit new skills
- Report broken/outdated skills
- Update existing skills from upstream
- Understand the repo's purpose and maintenance stance (README says "passive" maintenance)

**Recommendation:** Create `CONTRIBUTING.md` covering: skill submission format, metadata requirements, categorization rules, and testing expectations.

---

### P1-01: Complex Skill — claude-scientific-skills (14MB, 861 files)

**Location:** `skills_all/claude-scientific-skills/`  
**Evidence:**
- 144 Python scripts including a 1276-line document processor (`document-skills/docx/scripts/document.py`)
- 128+ scientific skills spanning bioinformatics, cheminformatics, proteomics, clinical research, medical imaging
- No `metadata.json` at the top level — invisible to `skills_categorized/` navigation
- No internal README explaining how to navigate or maintain this collection
- Largest sub-repository in the entire project

**Key files with concentrated expertise:**
| File | Lines | Domain |
|------|-------|--------|
| `scientific-skills/document-skills/docx/scripts/document.py` | 1276 | OOXML document manipulation |
| `scientific-skills/document-skills/pptx/scripts/inventory.py` | 1020 | PowerPoint generation |
| `scientific-skills/document-skills/pptx/ooxml/scripts/validation/base.py` | 966 | OOXML validation |
| `scientific-skills/exploratory-data-analysis/scripts/eda_analyzer.py` | 547 | Statistical EDA |
| `scientific-skills/statistical-analysis/scripts/assumption_checks.py` | 539 | Statistical testing |
| `scientific-skills/torch_geometric/scripts/create_gnn_template.py` | 529 | Graph neural networks |

**Impact:** This is effectively a sub-repo requiring deep scientific computing knowledge. A maintainer unfamiliar with OOXML, bioinformatics databases, or GNN architectures cannot effectively review or fix issues here.

**Recommendation:** Identify original upstream authors via metadata; document which skills are mirrored vs. modified; add a maintenance README.

---

### P1-02: Complex Skill — claude-win11-speckit-update-skill (248 files, 37 PS1 scripts)

**Location:** `skills_all/claude-win11-speckit-update-skill/`  
**Evidence:**
- 37 PowerShell scripts totaling 7,988 lines
- `update-orchestrator.ps1` alone is 1,059 lines with a 15-step workflow
- Has its own `.github/` with CI workflows (PR validation, fingerprint updates)
- Includes PRDs, release plans, spec docs (up to 1,805 lines)
- Entirely Windows/PowerShell domain — specialized expertise required

**Key complex files:**
| File | Lines | Risk |
|------|-------|------|
| `skills/speckit-updater/scripts/update-orchestrator.ps1` | 1059 | Main orchestrator, 15-step workflow |
| `.specify/scripts/powershell/update-agent-context.ps1` | 437 | Multi-agent context management |
| `skills/speckit-updater/scripts/helpers/Invoke-PreUpdateValidation.ps1` | 345 | Pre-update validation logic |
| `skills/speckit-updater/scripts/generate-fingerprints.ps1` | 299 | File fingerprinting for change detection |

**Impact:** This skill has more CI/CD sophistication than the entire parent repo. It's essentially an independent project with no documented connection to the parent's maintenance workflow.

**Recommendation:** Document this as a vendored dependency rather than an integrated skill; link to upstream repo for updates.

---

### P1-03: Undocumented Duplication Between skills_all and skills_categorized

**Location:** `skills_all/` (27MB) and `skills_categorized/` (31MB)  
**Evidence:**
- `skills_all/` contains 415 skill directories
- `skills_categorized/` contains 1,444 entries across 60 categories
- Same skill content duplicated across multiple categories (e.g., `frontend-design` appears in 20 categories, `api-design-principles` in 19)
- **15 skills exist in `skills_all/` but not in `skills_categorized/`**: `aws-skills`, `brand-guidelines`, `claude-scientific-skills`, `claude-win11-speckit-update-skill`, `claudisms`, `ffuf-claude-skill`, `internal-comms`, `ios-simulator-skill`, `notebooklm-skill`, `playwright-skill`, `superpowers-lab`, `template-skill`, `theme-factory`, `webapp-testing`, `xlsx`
- No documented logic for how/why skills are categorized
- No script or config file that drives categorization

**Missing from categorized view:**
```
skills_all/aws-skills                      # AWS CDK, serverless, cost ops
skills_all/claude-scientific-skills         # 128+ scientific skills
skills_all/claude-win11-speckit-update-skill # Windows update orchestrator
skills_all/ios-simulator-skill              # iOS testing automation
skills_all/notebooklm-skill                 # NotebookLM browser automation
skills_all/playwright-skill                 # Browser testing
skills_all/webapp-testing                   # Web app testing
```

**Impact:** Users browsing by category miss 15 skills including some of the most complex ones. Maintainers updating skills in `skills_all/` may forget to propagate to all categories.

**Recommendation:**
1. Create a `categorization.yaml` config mapping each skill to its categories
2. Generate `skills_categorized/` from `skills_all/` + config via a build script
3. Add CI check that validates `skills_categorized/` matches config

---

### P1-04: Complex Executable Skills Without Dependency Documentation

**Location:** Multiple skills  
**Evidence:** Only 3 of 415 skills have dependency files:
- `skills_all/notebooklm-skill/requirements.txt` (patchright, python-dotenv)
- `skills_all/ios-simulator-skill/pyproject.toml` (Python 3.12, ruff, black)
- `skills_all/playwright-skill/skills/playwright-skill/package.json`

Yet many skills reference external tools (e.g., `ffuf-claude-skill` requires Go/ffuf, `claude-scientific-skills` references rdkit, biopython, pytorch, etc.).

**Impact:** Users cannot determine what's needed to use a skill without reading the entire SKILL.md. No automated way to validate skill dependencies.

**Recommendation:** Add a `dependencies` field to `metadata.json` schema; validate during CI.

---

### P2-01: Inconsistent Skill Structure — 15 Skills Missing metadata.json

**Location:** `skills_all/` — 15 skill directories  
**Files affected:**
```
skills_all/aws-skills/
skills_all/brand-guidelines/
skills_all/claude-scientific-skills/
skills_all/claude-win11-speckit-update-skill/
skills_all/claudisms/
skills_all/ffuf-claude-skill/
skills_all/internal-comms/
skills_all/ios-simulator-skill/
skills_all/notebooklm-skill/
skills_all/playwright-skill/
skills_all/superpowers-lab/
skills_all/template-skill/
skills_all/theme-factory/
skills_all/webapp-testing/
skills_all/xlsx/
```

**Evidence:** 400 of 415 skills have `metadata.json`, but 15 do not. These 15 are disproportionately the most complex skills (multi-file, with scripts and nested directories).

**Impact:** Metadata-driven tooling (docs site, search, marketplace integration) cannot index these skills.

**Recommendation:** Generate `metadata.json` for all skills. Consider this a schema validation check in CI.

---

### P2-02: No CI/CD Pipeline

**Location:** `.github/workflows/` — does not exist at repo root  
**Evidence:** While individual skills carry their own CI (e.g., `claude-win11-speckit-update-skill/.github/workflows/`), the aggregation repo has zero validation workflows.

**Impact:** No automated checks for:
- Missing SKILL.md files
- Invalid metadata.json schema
- Broken internal links
- Stale categorization
- Duplicate content drift between `skills_all/` and `skills_categorized/`

**Recommendation:** Add GitHub Actions workflow that:
1. Validates every skill has SKILL.md + metadata.json
2. Checks metadata.json schema
3. Verifies skills_categorized is consistent with skills_all
4. Reports stats (skill count, category distribution)

---

### P2-03: Extremely Large SKILL.md Files — Monolithic Knowledge

**Location:** Multiple skills  
**Evidence:**
| Skill | Lines | Description |
|-------|-------|-------------|
| `treatment-plans/SKILL.md` | 1536 | Medical treatment plan generation |
| `github-project-management/SKILL.md` | 1277 | GitHub project management |
| `pair-programming/SKILL.md` | 1202 | Pair programming protocols |
| `hooks-automation/SKILL.md` | 1201 | Claude hooks automation |
| `design-by-contract/SKILL.md` | 1196 | Design by contract patterns |
| `flow-nexus-platform/SKILL.md` | 1157 | Flow Nexus platform |
| `github-code-review/SKILL.md` | 1140 | Code review workflows |
| `project-planning/SKILL.md` | 1121 | Project planning |
| `sparc-methodology/SKILL.md` | 1115 | SPARC methodology |

These are large monolithic documents with no modularization. `treatment-plans/SKILL.md` alone has 116 headings and 58 code blocks.

**Impact:** Large SKILL.md files consume excessive context window when loaded by Claude. Domain-specific sections are entangled. Hard to maintain and review.

**Recommendation:** For skills over 500 lines, split into `SKILL.md` (overview + triggers) + `references/` directory with detailed content.

---

### P2-04: Categorization Over-Saturation — 88 Skills in 5+ Categories

**Location:** `skills_categorized/`  
**Evidence:** 88 unique skill names appear in more than 5 categories. Top examples:
- `frontend-design`: 20 categories
- `fastapi-templates`: 20 categories
- `github-actions-templates`: 19 categories
- `api-design-principles`: 19 categories

**Impact:** Over-categorization reduces the value of categories as a navigation tool. Users see the same skill everywhere. No documented criteria for multi-category assignment.

**Recommendation:** Establish a maximum of 3-5 categories per skill; document the rationale for each assignment.

---

### P2-05: Template Skill Is a Stub

**Location:** `skills_all/template-skill/SKILL.md` (6 lines)  
**Evidence:**
```markdown
---
name: template-skill
description: Replace with description of the skill and when Claude should use it.
---

# Insert instructions below
```

**Impact:** New contributors looking for a template find a near-empty file. The template should demonstrate the expected structure, frontmatter schema, and best practices.

**Recommendation:** Expand template to include: frontmatter example, section structure (When to Use, Key Features, Configuration, Examples, Troubleshooting), and a metadata.json example.

---

### P3-01: Static Docs Site — Undocumented Build Process

**Location:** `docs/` (880KB)  
**Evidence:** The `docs/` directory contains a static site (`index.html`, `pages/*.md`) that powers `microck.github.io/ordinary-claude-skills`. There is no documented build script, no docs build config, and no reference to how pages are generated.

**Impact:** If the docs site needs updating, only Microck knows the process.

**Recommendation:** Document the docs build process in a `docs/README.md` or the main README. Add a Makefile or npm script.

---

### P3-02: License Ambiguity

**Location:** `LICENSE` (root), individual skill directories  
**Evidence:**
- Root LICENSE: MIT, Copyright (c) 2025 Microck
- README states: "i did not write most of these. i just collected them."
- Only 8 of 415 skills have their own LICENSE file
- Skills come from diverse upstream repos with potentially different licenses

**Impact:** Legal risk if upstream skills use non-MIT licenses. The blanket MIT claim may not apply to all content.

**Recommendation:** Audit upstream licenses; add license field to metadata.json; consider a `LICENSES/` directory with per-skill attribution.

---

### Immediate (Week 1) — Reduce P0 Risk
1. **Create `CONTRIBUTING.md`** with skill submission guidelines, structure requirements, and review process
2. **Create `CODEOWNERS`** with at least 2 maintainers listed
3. **Create `MAINTAINERS.md`** documenting the repo's purpose, update workflow, and categorization logic
4. **Recruit 1-2 co-maintainers** with visibility into the repo's operations

### Short-term (Month 1) — Address P1 Risks
5. **Create `categorization.yaml`** mapping each skill to its categories with rationale
6. **Write a build script** that generates `skills_categorized/` from `skills_all/` + config
7. **Add `metadata.json`** for the 15 missing skills
8. **Document the 4 complex multi-file skills** (scientific-skills, win11-speckit, ios-simulator, notebooklm) with architecture docs
9. **Audit the 15 uncategorized skills** and add them to appropriate categories

### Medium-term (Quarter 1) — Address P2 Risks
10. **Add CI/CD pipeline** with skill validation, metadata schema checks, and categorization consistency
11. **Refactor large SKILL.md files** (>500 lines) into modular structures with reference files
12. **Standardize the template skill** with a full example including all expected sections
13. **Add dependency declarations** to metadata.json schema
14. **Reduce over-categorization** — cap at 5 categories per skill

### Long-term (Ongoing)
15. **License audit** — verify all upstream licenses and document per-skill
16. **Document docs build process** for the static site
17. **Establish upstream sync process** — document how to pull updates from source repos
18. **Stale skill detection** — add metadata field for last-verified date; flag skills older than 6 months

---

## Repository Statistics Summary

| Metric | Value |
|--------|-------|
| Total contributors | 1 |
| Total commits | 1 |
| Skill directories (skills_all) | 415 |
| SKILL.md files | 551 |
| metadata.json files | 400 (15 missing) |
| Categorized entries (skills_categorized) | 1,444 |
| Categories | 60 |
| Code files (.py, .js, .ts, .ps1, .sh) | 201 |
| Largest skill (claude-scientific-skills) | 14MB, 861 files |
| Largest SKILL.md (examples.md) | 2,668 lines |
| Skills missing from categorized view | 15 |
| Skills in >5 categories | 88 |
| Repo-level CI/CD workflows | 0 |
| Repo-level CONTRIBUTING.md | Missing |
| Repo-level CODEOWNERS | Missing |
